post/2023-07-11-oneapi_profiling/index.md

```diff
@@ -0,0 +1,215 @@
+
+function kernel(a, b, c)
+    i = get_global_id()
+    @inbounds c[i] = a[i] + b[i]
+    return
+end
+
+function vadd(a, b)
+    d_a = oneArray(a)
+    d_b = oneArray(b)
+    d_c = similar(d_a)
+
+    @oneapi items=size(d_c) kernel(d_a, d_b, d_c)
+    Array(d_c)
+end
+
+function main(N=256)
+    a = round.(rand(Float32, N) * 100)
+    b = round.(rand(Float32, N) * 100)
+    c = vadd(a, b)
+end
+main()
+```
+
+We've tweaked this example to make it more suited for profiling: We've enclosed
+the main application in a function for performance improvement, and we've
+increased the array sizes to make the GPU work harder.
+
+There are several ways to profile this application. We'll start by demonstrating
+the command-line interface:
+
+```
+$ vtune -collect gpu-offload julia vadd.jl
+
+vtune: Collection started.
+vtune: Collection stopped.
+
+vtune: Using result path `/home/tim/Julia/pkg/oneAPI/r000gh'
+    GPU Time: 0.002s
+EU Array Stalled/Idle: 100.0% of Elapsed time with GPU busy
+ | The percentage of time when the EUs were stalled or idle is high, which has a
+ | negative impact on compute-bound applications.
+FPU Utilization: 0.0% of Elapsed time with GPU busy
+...
+```
+
+This will run the application, and collect a number of GPU-related metrics. A
+summary is shown in the terminal, and a more detailed report will be written
+to a directory in the current working directory. You can open that report with
+the graphical user interface, possibly even on a different machine:
+
+```
+$ vtune-gui r000gh
+```
+
+
+## Instrumenting the application
+
+The trace we just collected includes the time spent compiling our application,
+making it difficult to analyze what is happening. To refine the trace, we can
+instrument our application with Intel's Instrumentation and Tracing Technology
+(ITT) APIs:
+
+- only start the profiler when we're running code of interest;
+- add markers to the trace to indicate what is happening.
+
+We can interface with the ITT APIs using the IntelITT.jl package. Let's update
+our example:
+
+```julia
+using oneAPI, IntelITT
+
+# ...
+
+function main(N=256)
+    a = round.(rand(Float32, N) * 100)
+    b = round.(rand(Float32, N) * 100)
+    c = IntelITT.@range "vadd" oneAPI.@sync vadd(a, b)
+end
+
+# warm-up
+main()
+
+# actual profile
+IntelITT.@profile main()
+```
+
+Here, the `IntelITT.@profile` macro will start and stop the profiler, so we
+should launch VTune with the `-start-paused` option:
+
+```
+$ vtune -collect gpu-offload -start-paused julia vadd.jl
+```
+
+In the GUI, we can now clearly see a nicely packed stream of API calls, grouped
+under the `vadd` range we added. Note that because API calls are asynchronous,
+i.e. they return immediately before the GPU has executed them, I grouped them
+under a `oneAPI.@sync` call so that the range not only captures the time spent
+on the CPU, but also the time spent on the GPU. This may not be wanted for your
+application.
+
+![VTune timeline](vtune_timeline.png)
+
+
```

Comment:
Maybe add the vtune command line for gpu-hotspot for this section matching the gpu-offload section above?

Reply:
I mean adding after line 156.